ambry-cloud/src/main/java/com.github.ambry.cloud/CloudStorageCompactor.java

```diff
@@ -91,14 +99,25 @@ public int compactPartitions() {
   /**
    * Purge the inactive blobs in the specified partition.
    * @param partitionPath the partition to compact.
-   * @return the number of blobs purged.
+   * @param cutoffTime the time at which a blob's active status should be evaluated.
+   * @return the number of blobs purged or found.
    */
-  public int compactPartition(String partitionPath) throws CloudStorageException {
-    List<CloudBlobMetadata> deadBlobs = cloudDestination.getDeadBlobs(partitionPath);
-    if (!testMode) {
-      return cloudDestination.purgeBlobs(deadBlobs);
-    } else {
-      return deadBlobs.size();
+  public int compactPartition(String partitionPath, long cutoffTime) throws CloudStorageException {
+    if (testMode) {
+      return cloudDestination.getDeadBlobCount(partitionPath, cutoffTime);
     }
+
+    int numDeadBlobs = 0;
+    // Iterate until returned list size < limit or time runs out
+    while (System.currentTimeMillis() < cutoffTime + compactionTimeLimitMs) {
+      List<CloudBlobMetadata> deadBlobs =
```

Comment:
do we need to sleep in between successive calls to getDeadBlobs, or do you feel that the cosmo's retry-with-backoff will be good enough for rate limiting this query?

Reply:
There is an expensive call to cloudDestination.purgeBlobs() between calls to getDeadBlobs, and that should provide more than enough throttling between queries.